templates/flasher.lua

```diff
@@ -960,6 +961,80 @@ end
 
 
 
+-- Smart Erase. Erase an area in SPI-Flash with automatic choice of the optimal erase Commands.
+-- The start and end addresses must be aligned to sector boundaries as
+-- set by getEraseArea.
+function smart_erase(tPlugin, aAttr, ulEraseStart, ulEraseEnd, fnCallbackMessage, fnCallbackProgress)
+	local aulParameter = 
+	{
+		OPERATION_MODE_SmartErase,                     -- operation mode: smart_erase
+		aAttr.ulDeviceDesc,                            -- data block for the device description
+		ulEraseStart,
+		ulEraseEnd,
+	}
+	local ulValue = callFlasher(tPlugin, aAttr, aulParameter, fnCallbackMessage, fnCallbackProgress)
+	return ulValue == 0
+  end
+
+
+-----------------------------------------------------------------------------
+-- Erase an area with mart erase sizes:
+-- ulSize = 0xffffffff to erase from ulDeviceOffset to end of chip
+--
+-- OK:
+-- Area erased
+--
+-- Error messages:
+-- getEraseArea failed!
+-- Failed to erase the area! (Failure during smart_erase)
+-- Failed to erase the area! (isErased check failed)
+
+
+function smartEraseArea(tPlugin, aAttr, ulDeviceOffset, ulSize, fnCallbackMessage, fnCallbackProgress)
+	local fIsErased
+	local ulEndOffset
+	local ulEraseStart,ulEraseEnd
+
+	-- If length = 0xffffffff we get the erase area now in order to detect the flash size.
+	if ulSize == 0xffffffff then
+		ulEndOffset = ulSize
+		ulEraseStart,ulEraseEnd = getEraseArea(tPlugin, aAttr, ulDeviceOffset, ulEndOffset, fnCallbackMessage, fnCallbackProgress)
+		if not (ulEraseStart and ulEraseEnd) then
+			return false, "getEraseArea failed!"
+		end
+		
+		ulEndOffset = ulEraseEnd
+	else
+		ulEndOffset = ulDeviceOffset + ulSize
+	end
+	
+	print(string.format("Area:  [0x%08x, 0x%08x[", ulDeviceOffset, ulEndOffset))
+	print("Checking if the area is already empty")
+	fIsErased = isErased(tPlugin, aAttr, ulDeviceOffset, ulEndOffset, fnCallbackMessage, fnCallbackProgress)
```

Comment:
We might remove this check, because the smart erase already checks if the blocks are empty.

Reply:
I didn't include this initially, however the new smart erase operation will fall back to the normal erase operation under various conditions:
1. We are not erasing the SPI-Flash but one of the other memories
2. There are less than 2 erase operations found in the erase operations array
3. The largest erase command is so large, we run out of memory to keep track of it (highly unlikely)

Point 2. is currently the case for most netX, as far as I can tell. This is because they get their flash information from the flasher-included table which currently only supports one sector erase and one page erase command.

I'd be open to removing this but I'm not certain if we should remove it, given the reason stated above.
What do you think?

Reply:
I'm sorry, I didn't think of the fallback. It makes sense to keep the initial check. Maybe this can be optimized later on.

Reply:
About the fallback in case 2 (and 3?): Even if there's only one size of erase block available, don't we want to erase only those blocks that are not already empty?

Reply:
On the one hand, it does not matter, since smart erase currently only works with SPI flash, on the other hand, we generally want to fall back to normal erase automatically if smart erase is not possible for some reason, so we'll keep the is_erased check.